src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs

```diff
@@ -85,13 +85,24 @@ pub(super) fn relate_type_and_user_type<'tcx>(
     // variance to get the right relationship.
     let v1 = ty::Contravariant.xform(v);
 
-    TypeRelating::new(
+    let mut type_relating = TypeRelating::new(
         infcx.tcx,
         NllTypeRelatingDelegate::new(infcx, borrowck_context, locations, category),
         v1,
         b_variables,
-    ).relate(&b_value, &a)?;
-    Ok(())
+    );
+    type_relating.relate(&b_value, &a)?;
```

Comment:
**type_relating.relate(&b_value, &a)?;**

Can you explain to me what this line does ?

Reply:
It checks that `a` and `b_value` are the same type apart from lifetimes, if they're not it returns to the caller. and adds any outlives constraints that are required for `a` being a subtype of `b`, or `b` being a subtype of `a` or `b` being the same type as `a` to the `borrowck_context` (if it's `Some`, which it will be during the NLL type check). Finally it works out what any type variables in `b` should be based on `a`.

So if `b` is `&mut '_1 &'_2 _`, `a` is `&mut '_3 &'_4 i32 and `v1` is Covariant, the constraints added are `'_1: '_3`, `'_2: '_4` and `'_4: '_2` (the last constraint comes from `&mut T` being invariant in `T`), and the type variable is inferred to be `i32`.

Reply:
@matthewjasper Thanks a lot for the explanation. Really appreciate it.